File: nflog_sniffer.py
#!/usr/bin/python
#
# Written by Andreas Jaggi <andreas.jaggi@waterwave.ch> in December 2015
#

from socket import AF_INET, AF_INET6, inet_ntop

import nflog

from dpkt import ip, dns

def callback(payload):
	pkt = ip.IP(payload.get_data())
	if pkt.p == ip.IP_PROTO_UDP:
		pack = dns.DNS(pkt.udp.data)
		if pack.qr == dns.DNS_R:
			for rr in pack.an:
				if rr.type == dns.DNS_A:
					print "answer(%s)[%d]: %s" % (rr.name, rr.ttl, inet_ntop(AF_INET,rr.ip))
				if rr.type == dns.DNS_AAAA:
					print "answer(%s)[%d]: %s" % (rr.name, rr.ttl, inet_ntop(AF_INET6,rr.ip6))
				if rr.type == dns.DNS_CNAME:
					print "answer(%s)[%d]: %s" % (rr.name, rr.ttl, rr.cname)
		

def main():
	l = nflog.log()
	l.set_callback(callback)
	l.fast_open(123,AF_INET)
	try:
		l.try_run()
	except KeyboardInterrupt, e:
		print "interrupted, terminating..."
	l.unbind(AF_INET)
	l.close()
	

if __name__ == "__main__":
	main()

File: nflog_sniffer.pl
#!/usr/bin/perl
#
# Written by Andreas Jaggi <andreas.jaggi@waterwave.ch> in December 2015
#

use strict;
use warnings;

use nflog;
use NetPacket::IP qw(IP_PROTO_UDP);
use NetPacket::UDP;
use Socket qw(AF_INET AF_INET6);
use Net::DNS::Packet;

my $l;

sub callback {
	my ($payload) = @_;
	if ($payload) {
		my $ip_obj = NetPacket::IP->decode($payload->get_data());
		if ($ip_obj->{proto} == IP_PROTO_UDP) {
			my $packet = new Net::DNS::Packet(\NetPacket::UDP->decode($ip_obj->{data})->{data});
			if($packet->header->qr){
				foreach my $answer ( $packet->answer ) {
					print "answer(".$answer->type.")[".$answer->ttl."]: ".$answer->name." -> ".$answer->rdstring."\n";
				}
			}
		}
	}
	return 0;
}

sub cleanup {
	$l->unbind(AF_INET);
	$l->close();
}

sub main {
	$l = new nflog::log();
	$SIG{INT} =  "cleanup";
	$l->set_callback(\&callback);
	$l->fast_open(123,AF_INET);
	$l->try_run();
}

main();
